File: 04/04.pl
#!/usr/bin/perl
# Day 4: Camp Cleanup
# https://adventofcode.com/2022/day/4
# J.Christensen 04Dec2022

use v5.24;
use strict;
use warnings;
use utf8;
use List::Util qw(min max);
binmode STDOUT, ':encoding(utf8)';
binmode STDERR, ':encoding(utf8)';
say "\n---- PERL $0 ----";

# read the input, strip newlines
chomp(my @lines = <<>>);

#---- PART ONE ----#
my $total = 0;
foreach (@lines) {
    my ($f1, $f2, $f3, $f4) = split/,|-/;
    my $min1 = min($f1, $f2);
    my $max1 = max($f1, $f2);
    my $min2 = min($f3, $f4);
    my $max2 = max($f3, $f4);
    if ( (($min1 >= $min2) && ($max1 <= $max2)) || (($min2 >= $min1) && ($max2 <= $max1)) ) {
        $total++;
    }
}
say "Part 1 total $total.";

#---- PART TWO ----#
$total = 0;
foreach (@lines) {
    my ($f1, $f2, $f3, $f4) = split/,|-/;
    my $min1 = min($f1, $f2);
    my $max1 = max($f1, $f2);
    my $min2 = min($f3, $f4);
    my $max2 = max($f3, $f4);
    if ( (($max1 >= $min2) && ($max1 <= $max2)) || (($max2 >= $min1) && ($max2 <= $max1)) ) {
        $total++;
    }
}
say "Part 2 total $total.";

File: 04/04.py
#!/usr/bin/python3
# Day 4: Camp Cleanup
# https://adventofcode.com/2022/day/4
# J.Christensen 04Dec2022

import os
import re
import sys

def main() -> None:
    print(f"\n---- PYTHON {os.path.basename(__file__)} ----")

    # read the input, strip leading/trailing whitespace including newlines
    lines = sys.stdin.readlines()
    for i in range(len(lines)):
        lines[i] = lines[i].strip()

    # delimiters to split input
    reDelim = re.compile(r',|-')

    #---- PART ONE ----#
    total = 0
    # parse the fields; convert to integer
    for l in lines:
        f = re.split(reDelim, l)
        for i in range(len(f)):
            f[i] = int(f[i])
        min1 = min(f[0], f[1])
        max1 = max(f[0], f[1])
        min2 = min(f[2], f[3])
        max2 = max(f[2], f[3])
        if ( ((min1 >= min2) and (max1 <= max2)) or ((min2 >= min1) and (max2 <= max1)) ):
            total += 1
    print(f'Part 1 total {total}.')

    #---- PART TWO ----#
    total = 0
    # parse the fields; convert to integer
    for l in lines:
        f = re.split(reDelim, l)
        for i in range(len(f)):
            f[i] = int(f[i])
        min1 = min(f[0], f[1])
        max1 = max(f[0], f[1])
        min2 = min(f[2], f[3])
        max2 = max(f[2], f[3])
        if ( ((max1 >= min2) and (max1 <= max2)) or ((max2 >= min1) and (max2 <= max1)) ):
            total += 1
    print(f'Part 2 total {total}.')

if __name__ == '__main__':
    main()
